sdk/inc/azure/iot/az_iot_common.h

```diff
@@ -61,17 +61,112 @@ typedef enum
+    uint32_t current_property_index;
+  } _internal;
+} az_iot_message_properties;
+
+/**
+ * @brief Initializes the Telemetry or C2D properties.
+ *
+ * @note The properties init API will not encode properties. In order to support
+ *       the following characters, they must be percent-encoded (RFC3986) as follows:
+ *          `/` : `%2F`
+ *          `%` : `%25`
+ *          `#` : `%23`
+ *          `&` : `%26`
+ *       Only these characters would have to be encoded. If you would like to avoid the need to
+ *       encode the names/values, avoid using these characters in names and values.
+ *
+ * @param[in] properties The #az_iot_message_properties to initialize
+ * @param[in] buffer Can either be an empty #az_span or an #az_span containing properly formatted
+ *                   (with above mentioned characters encoded if applicable) properties with the
+ *                   following format: {key}={value}&{key}={value}.
+ * @param[in] written_length The length of the properly formatted properties already initialized
+ * within the buffer. If the \p buffer is empty (uninitialized), this should be 0.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_message_properties_init(
+    az_iot_message_properties* properties,
+    az_span buffer,
+    int32_t written_length);
+
+/**
+ * @brief Appends a key-value property to the list of properties.
+ *
+ * @note The properties append API will not encode properties. In order to support
+ *       the following characters, they must be percent-encoded (RFC3986) as follows:
+ *          `/` : `%2F`
+ *          `%` : `%25`
+ *          `#` : `%23`
+ *          `&` : `%26`
+ *       Only these characters would have to be encoded. If you would like to avoid the need to
+ *       encode the names/values, avoid using these characters in names and values.
+ *
+ * @param[in] properties The #az_iot_message_properties to use for this call
+ * @param[in] name The name of the property.
+ * @param[in] value The value of the property.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_message_properties_append(
+    az_iot_message_properties* properties,
+    az_span name,
+    az_span value);
+
+/**
+ * @brief Finds the value of a property.
+ * @remark This will return the first value of the property with the given name if multiple
+ * properties with the same key exist.
+ *
+ * @param[in] properties The #az_iot_message_properties to use for this call
+ * @param[in] name The name of the property.
+ * @param[out] out_value An #az_span containing the value of the property.
+ * @return #az_result.
+ */
+AZ_NODISCARD az_result az_iot_message_properties_find(
+    az_iot_message_properties* properties,
+    az_span name,
+    az_span* out_value);
+
+/**
+ * @brief Iterates over the list of properties.
+ *
+ * @param[in] properties The #az_iot_message_properties to use for this call
+ * @param[out] out An #az_pair containing the key and the value of the next property.
+ * @return #az_result
```

Comment:
Outside the scope of this PR (please do this separately), but consider updating the `@return` docs to be more descriptive.

Follow the approach from core and add this line for any az_result return:
```text
@return An #az_result value indicating the result of the operation.
```

Followed by a list of relevant `@retval` results that the user would want to know, specific to this API.

https://github.com/Azure/azure-sdk-for-c/blob/332102c172b7b3b2788e1c46a44e572e04d4673a/sdk/inc/azure/core/az_json.h#L399-L408